File: Leetcode/49 - Group Anagrams/solution.ts
function groupAnagrams(strs: string[]): string[][] {
    const anagrams: Map<string, string[]> = new Map();

    for (const str of strs) {
        const key = str.split('').sort().join('');
        if (!anagrams.has(key)) {
            anagrams.set(key, []);
        }
        anagrams.get(key)!.push(str);
    }

    return Array.from(anagrams.values());
};

File: Leetcode/49 - Group Anagrams/solution.cpp
vector<vector<string>> groupAnagrams(vector<string>& strs) {
    unordered_map<string, vector<string>> anagrams;
    for (string& str: strs) {
        string s = str;
        sort(s.begin(), s.end());
        anagrams[s].push_back(str);
    }

    vector<vector<string>> group;
    for (auto& pair: anagrams) {
        group.push_back(pair.second);
    }

    return group;
}